File: backend/semgrep_rules/tests/python/kisa_sw_02_code_injection.py
def unsafe_eval():
    code = input(
        "code: "
    )

    # ruleid: kisa.sw02.python.external-input-to-eval-exec
    return eval(
        code
    )


def unsafe_exec():
    script = input(
        "script: "
    )

    # ruleid: kisa.sw02.python.external-input-to-eval-exec
    exec(
        script
    )


def unsafe_compile():
    source = input(
        "source: "
    )

    # ruleid: kisa.sw02.python.external-input-to-eval-exec
    compiled = compile(
        source,
        "<user-input>",
        "exec",
    )

    return compiled


def unsafe_django_get(
    request
):
    expression = (
        request.GET.get(
            "expression"
        )
    )

    # ruleid: kisa.sw02.python.external-input-to-eval-exec
    return eval(
        expression
    )


def unsafe_django_post(
    request
):
    script = (
        request.POST.get(
            "script"
        )
    )

    # ruleid: kisa.sw02.python.external-input-to-eval-exec
    exec(
        script
    )


def safe_value():
    value = input(
        "value: "
    )

    # ok: kisa.sw02.python.external-input-to-eval-exec
    return {
        "value":
            value
    }


def safe_mapping():
    action = input(
        "action: "
    )

    handlers = {
        "list":
            lambda: "list",

        "detail":
            lambda: "detail",
    }

    # ok: kisa.sw02.python.external-input-to-eval-exec
    handler = (
        handlers.get(
            action
        )
    )

    if handler is None:
        return None

    return handler()

File: backend/semgrep_rules/tests/javascript/kisa_sw_02_code_injection.js
function unsafeEval(req) {
  const code = req.query.code;

  // ruleid: kisa.sw02.javascript.external-input-to-dynamic-code
  return eval(code);
}


function unsafeFunction(req) {
  const body = req.body.script;

  // ruleid: kisa.sw02.javascript.external-input-to-dynamic-code
  const fn = new Function(body);

  return fn();
}


function safeMapping(req) {
  const action = req.query.action;

  const handlers = {
    list: () => "list",
    detail: () => "detail",
  };

  // ok: kisa.sw02.javascript.external-input-to-dynamic-code
  return handlers[action]?.();
}
